src/System.Private.CoreLib/shared/System/Diagnostics/DebugProvider.cs

```diff
@@ -72,38 +73,38 @@ public virtual void WriteLine(string message)
 
         private sealed class DebugAssertException : Exception
         {
-            internal DebugAssertException(string stackTrace) :
+            internal DebugAssertException(string? stackTrace) :
                 base(Environment.NewLine + stackTrace)
             {
             }
 
-            internal DebugAssertException(string message, string stackTrace) :
+            internal DebugAssertException(string? message, string? stackTrace) :
                 base(message + Environment.NewLine + Environment.NewLine + stackTrace)
             {
             }
 
-            internal DebugAssertException(string message, string detailMessage, string stackTrace) :
+            internal DebugAssertException(string? message, string? detailMessage, string? stackTrace) :
                 base(message + Environment.NewLine + detailMessage + Environment.NewLine + Environment.NewLine + stackTrace)
             {
             }
         }
 
         private bool _needIndent = true;
 
-        private string _indentString;
+        private string? _indentString;
 
         private string GetIndentString()
         {
             int indentCount = Debug.IndentSize * Debug.IndentLevel;
             if (_indentString?.Length == indentCount)
             {
-                return _indentString;
+                return _indentString!;
```

Comment:
This is an interesting case.  The compiler is correct here, in that indentCount could be 0.  And if indentCount is 0, and if _indentString is null, then _indentString?.Length == indentCount is true, and we end up returning a null string, which probably wasn't intended.  In practice, it doesn't matter, because the only call site here just uses GetIndentString in a string.Concat, which will treat a null the same as a string.Empty.

The right fix here isn't to `!` on `_indentString, but rather to declare `GetIndentString` as returning `string?`, since it can actually return null and callers need to be ok with that.

Reply:
I tested 
```
        static void Main(string[] args)
        {
            string _indentString = null;
            int indentCount = 0;
            if (_indentString?.Length == indentCount)
            {
                Console.WriteLine("null.length == 0");
            }
            else
            {
                Console.WriteLine("Null.length not equal to 0");
            }
            Console.Read();
        }
```
Here it enters in else condition rather than if , so it wont return null

Reply:
Interesting; it generates different code than I thought it did for this, changing the condition to be the equivalent of `if (_indentString != null && _indentString.Length == indentCount)`.

Are you sure the `!` on _indentString is necessary?

Reply:
if we change condition to ``` if (_indentString != null && _indentString.Length == indentCount)```
then we don't need `!` .

if we don't want to change the if condtion we have to add ```!```
which path should we take here ?

Reply:
Are you getting the error when building corelib at the command-line?

Reply:
yes I am getting error while building corelib from the command line
```
shared\System\Diagnostics\DebugProvider.cs(101,24): error CS8603: Possible null reference return. [C:\git\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
```

Reply:
Ok.  Use `!` and add a `// TODO-NULLABLE: https://github.com/dotnet/roslyn/issues/34942` comment.